### PR TITLE
Add manifest for TigerVNC

### PR DESCRIPTION
**Pull request by acperkins**
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. [Contacted upstream](https://github.com/TigerVNC/tigervnc/issues/1590).
- [X] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. [Pull request](https://github.com/TigerVNC/tigervnc/pull/1592).

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: org.tigervnc.TigerVNC.yml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        url: https://www.fltk.org/pub/fltk/1.3.8/fltk-1.3.8-source.tar.gz
+        sha256: f3c1102b07eb0e7a50538f9fc9037c18387165bc70d4b626e94ab725b9d4d1bf
+  - name: pam
+    buildsystem: simple
+    sources:
+      - type: archive
+        url: https://github.com/linux-pam/linux-pam/releases/download/v1.5.2/Linux-PAM-1.5.2.tar.xz
+        sha256: e4ec7131a91da44512574268f493c6d8ca105c87091691b8e9b56ca685d4f94d
+    build-commands:
+      - ./configure --prefix=$FLATPAK_DEST
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml:22]: this should be `buildsystem: autotools`

File: org.tigervnc.TigerVNC.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  - name: appdata
+    buildsystem: simple
+    sources:
+      - type: file
+        url: https://raw.githubusercontent.com/TigerVNC/tigervnc/704eff041c3e5dd0fb712fa76774b8d3f6dd31f9/vncviewer/org.tigervnc.vncviewer.metainfo.xml.in
+        dest-filename: org.tigervnc.TigerVNC.metainfo.xml
+        sha256: 743a7565b95490ede10eae15259fab432fd1c385adece44c455553f7faf21bb2
+    build-commands:
+      - install -Dm644 org.tigervnc.TigerVNC.metainfo.xml $FLATPAK_DEST/share/metainfo/org.tigervnc.TigerVNC.metainfo.xml
+finish-args:
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml:43]: these two are usually above the `modules`

File: org.tigervnc.TigerVNC.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
+app-id: org.tigervnc.TigerVNC
+runtime: org.freedesktop.Platform
+runtime-version: '22.08'
+sdk: org.freedesktop.Sdk
+command: vncviewer
+rename-icon: tigervnc
+rename-desktop-file: vncviewer.desktop
+modules:
+  - name: fltk
+    buildsystem: cmake
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml:10]:
```suggestion
    buildsystem: cmake-ninja
```

File: org.tigervnc.TigerVNC.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      - type: archive
+        url: https://github.com/linux-pam/linux-pam/releases/download/v1.5.2/Linux-PAM-1.5.2.tar.xz
+        sha256: e4ec7131a91da44512574268f493c6d8ca105c87091691b8e9b56ca685d4f94d
+    build-commands:
+      - ./configure --prefix=$FLATPAK_DEST
+      - make
+      - make install
+      - cp -r libpam/include/* $FLATPAK_DEST/include
+  - name: vncviewer
+    buildsystem: cmake
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml:27]:
```suggestion
    buildsystem: cmake-ninja
```

File: org.tigervnc.TigerVNC.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      - "-DCMAKE_CXX_FLAGS=-lpng16"
+  - name: appdata
+    buildsystem: simple
+    sources:
+      - type: file
+        url: https://raw.githubusercontent.com/TigerVNC/tigervnc/704eff041c3e5dd0fb712fa76774b8d3f6dd31f9/vncviewer/org.tigervnc.vncviewer.metainfo.xml.in
+        dest-filename: org.tigervnc.TigerVNC.metainfo.xml
+        sha256: 743a7565b95490ede10eae15259fab432fd1c385adece44c455553f7faf21bb2
+    build-commands:
+      - install -Dm644 org.tigervnc.TigerVNC.metainfo.xml $FLATPAK_DEST/share/metainfo/org.tigervnc.TigerVNC.metainfo.xml
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml:42]:
```suggestion
    config-opts:
      - "-DCMAKE_CXX_FLAGS=-lpng16"
    post-install:
      - install -Dm644 org.tigervnc.TigerVNC.metainfo.xml $FLATPAK_DEST/share/metainfo/org.tigervnc.TigerVNC.metainfo.xml
    sources:
      - type: archive
        url: https://github.com/TigerVNC/tigervnc/archive/refs/tags/v1.12.0.tar.gz
        sha256: 7bbf660f315d3bbac370757dbabfb6853263d8b5380f59ded15b14daf97141d7
      - type: file
        url: https://raw.githubusercontent.com/TigerVNC/tigervnc/704eff041c3e5dd0fb712fa76774b8d3f6dd31f9/vncviewer/org.tigervnc.vncviewer.metainfo.xml.in
        dest-filename: org.tigervnc.TigerVNC.metainfo.xml
        sha256: 743a7565b95490ede10eae15259fab432fd1c385adece44c455553f7faf21bb2
```

File: org.tigervnc.TigerVNC.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
+app-id: org.tigervnc.TigerVNC
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml:1]: the upstream adddate uses `org.tigervnc.vncviewer`. It has to match one way or the other.

**Comment by hfiguiere** [comment]
I know it will fail because of the appstream file:

bot, build org.tigervnc.TigerVNC

File: org.tigervnc.TigerVNC.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  - --socket=x11
+cleanup:
+  - '/bin/fltk-config'
+  - '/bin/fluid'
+  - '/bin/vncconfig'
+  - '/bin/vncpasswd'
+  - '/bin/x0vncserver'
+  - '/etc'
+  - '/include'
+  - '/lib/debug'
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml:56]:
```suggestion
```

File: metainfo.patch [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+   <screenshots>
+     <screenshot type="default">
+       <caption>TigerVNC Viewer connection to a CentOS machine</caption>
+@@ -42,4 +44,52 @@
+       <image>https://raw.githubusercontent.com/TigerVNC/tigervnc/741d3edbfab65eda6f033078bc06347fe244ea6a/vncviewer/metainfo/tigervnc-connection-windows.jpg</image>
+     </screenshot>
+   </screenshots>
++
++  <releases>
++    <release type="stable" version="1.12.0" date="2021-11-10T07:51:00Z">
```
Comment (hfiguiere) [inline metainfo.patch:27]:
```suggestion
+    <release type="stable" version="1.12.0" date="2021-11-10">
```

File: metainfo.patch [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
++  <url type="bugtracker">https://github.com/TigerVNC/tigervnc/issues</url>
+   <description>
+     <p>
+       Virtual Network Computing (VNC) is a remote display system that allows you to view and interact with a virtual desktop environment running on another computer on the network.
+@@ -27,7 +29,7 @@
+     </p>
+   </description>
+ 
+-  <launchable type="desktop-id">vncviewer.desktop</launchable>
++  <launchable type="desktop-id">org.tigervnc.TigerVNC.desktop</launchable>
```
Comment (hfiguiere) [inline metainfo.patch:17]: FWIW, this is covered by the `rename-desktop-file`.

File: org.tigervnc.TigerVNC.yml [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    sources:
+      - type: file
+        url: https://raw.githubusercontent.com/TigerVNC/tigervnc/704eff041c3e5dd0fb712fa76774b8d3f6dd31f9/vncviewer/org.tigervnc.vncviewer.metainfo.xml.in
+        dest-filename: org.tigervnc.TigerVNC.metainfo.xml
+        sha256: 743a7565b95490ede10eae15259fab432fd1c385adece44c455553f7faf21bb2
+      - type: patch
+        path: metainfo.patch
+        strip-components: 0
+    build-commands:
+      - install -Dm644 org.tigervnc.TigerVNC.metainfo.xml $FLATPAK_DEST/share/metainfo/org.tigervnc.TigerVNC.metainfo.xml
```
Comment (hfiguiere) [inline org.tigervnc.TigerVNC.yml]:
```suggestion
    config-opts:
      - "-DCMAKE_CXX_FLAGS=-lpng16"
    post-install:
      - install -Dm644 org.tigervnc.TigerVNC.metainfo.xml $FLATPAK_DEST/share/metainfo/org.tigervnc.TigerVNC.metainfo.xml
    sources:
      - type: git
        url: https://github.com/TigerVNC/tigervnc.git
        tag: v1.12.0
        commit: f1f65c5904e6143cd242c0f1c53d7f8147dabd74
      - type: file
        url: https://raw.githubusercontent.com/TigerVNC/tigervnc/704eff041c3e5dd0fb712fa76774b8d3f6dd31f9/vncviewer/org.tigervnc.vncviewer.metainfo.xml.in
        dest-filename: org.tigervnc.TigerVNC.metainfo.xml
        sha256: 743a7565b95490ede10eae15259fab432fd1c385adece44c455553f7faf21bb2
      - type: patch
        path: metainfo.patch
        strip-components: 0
```

File: metainfo.patch [hunk 11]
```diff
@@ -0,0 +1,80 @@
+--- org.tigervnc.vncviewer.metainfo.xml.in	2023-01-31 21:29:23.779423487 +0000
++++ org.tigervnc.TigerVNC.metainfo.xml	2023-01-31 21:30:43.924526110 +0000
+@@ -7,12 +7,14 @@
+   notice are preserved. This file is offered as-is, without any warranty.
+ -->
+ <component type="desktop-application">
+-  <id>org.tigervnc.vncviewer</id>
++  <id>org.tigervnc.TigerVNC</id>
```
Comment (hfiguiere) [inline metainfo.patch:8]: The preference here would be that the upstream chosen app-id be used instead.

Comment (acperkins) [reply metainfo.patch:8]: Thanks, I'll change that now. I picked ".TigerVNC" to match the case of other apps on Flathub and to distinguish from other "vncviewer"s, but I suppose the reverse domain name does the latter.

Comment (hfiguiere) [reply metainfo.patch:8]: The name says `TigerVNC Viewer` so it's clear to the user

**Comment by acperkins** [comment]
I think I've made all of the recommended changes now. I have contacted upstream to check that they are happy for me to submit the application and use their domain in the appid.

File: org.tigervnc.vncviewer.yml [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    sources:
+      - type: git
+        url: https://github.com/TigerVNC/tigervnc.git
+        tag: v1.12.0
+        commit: f1f65c5904e6143cd242c0f1c53d7f8147dabd74
+  - name: appdata
+    buildsystem: simple
+    build-commands:
+      - install -Dm644 org.tigervnc.vncviewer.metainfo.xml $FLATPAK_DEST/share/metainfo/org.tigervnc.vncviewer.metainfo.xml
+    sources:
```
Comment (hfiguiere) [inline org.tigervnc.vncviewer.yml:64]:
```suggestion
    post-install:
      - install -Dm644 org.tigervnc.vncviewer.metainfo.xml $FLATPAK_DEST/share/metainfo/org.tigervnc.vncviewer.metainfo.xml
    sources:
      - type: git
        url: https://github.com/TigerVNC/tigervnc.git
        tag: v1.12.0
        commit: f1f65c5904e6143cd242c0f1c53d7f8147dabd74
```

**Comment by hfiguiere** [comment]
bot, build org.tigervnc.vncviewer

**Comment by acperkins** [comment]
I've made the `<releases>` tag smaller in metainfo, so it's easier to maintain until upstream adopt the changes.

**Comment by barthalion** [comment]
Is this good to go? It's still marked as draft.

**Comment by acperkins** [comment]
> Is this good to go? It's still marked as draft.

Yes, good to go now. Thanks.

**Pull request closed without merging**